### ISSUE TITLE
Boolean expression Evaluator

### ISSUE THREAD

**Issue by njrapidinnovation** [OpenZeppelin/openzeppelin-contracts#3661]
**🧐 Motivation**
Easy and Convenient way to evaluate a complex and dynamic boolean expression with precedence.

**📝 Details**

Provides method that can be used to evalute any Infix Notation Expression which

*  consist of 1 and 0 as values .
*  consist of a (&&) , o (||) , ( , ) as operators.
*  precedence is also considered for LOGICAL AND (&&) and LOGICAL OR (||)
*  function parameters are 
    *  String Array of 1 , 0 , o , a , ( , ) where o is Logical OR (||) and a is Logical AND (&&).
    *  noOfValues i.e count of 1 and 0 in array.
    *  noOfOperands i.e count of a, o, ( , ) in array.

Related to #3660

**Comment by Amxx** [comment]
Hello @njrapidinnovation 

Can you please explain what usecase you see for such a library ? Beyond the experimental aspect of "it can be done in solidity", I'm failing to see the actual point.

Keep in mind that it may be fun for you to build, but if we merge it, we are committing ourselves to maintaining it.

**Comment by njrapidinnovation** [comment]
Yeah sure @Amxx 

We can calculate a result of nested if else logic.

For example :-

```

    if(c1){
    }
    else{
       if(c1 || c2){
       }
       else{
           if(c3 && ( c4 || c5)){
           }
       }
    }

    where c(1...5) are all boolean expressions

    You can rewrite the above logic as

    (c1 || (c1 || c2) || (c3 && (c4 || c5)))

```

```
   
   Pratical Application:-
   
   * You can make a conditional NFT renting platform where a user can choose different
   criterias among the listed criterias to create an expression in order to rent a
   NFT

   * Like User Renting is Eligible until below expression returns true

   * (time < 25th decemeber || price of NFT < 100usd) && (weather == summer)   

   Or they can list out the criterias which the community want.   



  Basically using this you can build a platform in which services will work upon dynamic boolean expressions.

 
```

**Comment by njrapidinnovation** [comment]
Hello @Amxx 
Any thoughts on the above use case?

**Comment by Amxx** [comment]
Honestly I think its over complicated and not realistic. 

Anything like `weather == summer` would have to be encoded and the push into the array before getting processed..., and the other input would have to be validated somehow. It can possibly de done nicely, but again, I don't think this aligns with what we want to propose and maintain.

I'll quote @frangio here:
> However, these are experimental projects and OpenZeppelin Contracts is not the place for them. We try to focus on things that have proven value, and that will be generally useful.

That doesn't mean this shouldn't be studied further, just here is not the right place.
